[vcpkg] Registries MVP

strega-nil opened the pull request:
This adds everything except for git registries, which will be implemented later.

Note: this is a large change. You may want to look at each commit in turn.

See <https://github.com/strega-nil/vcpkg-example/tree/trunk> for an example of using registries :) (ignore the actual C++, that's not really what's important... no guarantees on it compiling, for example)

ras0219-msft committed:
[vcpkg] Enable Optional<T> -> Optional<T&> conversion

Also move common_projection() to Util:: to be with its friends

ras0219-msft committed:
[vcpkg] Load, parse, and extract configuration from manifests in VcpkgPaths

This PR adds support for $.configuration in manifest files, which can either be an immediate object or the string "file" (which will search recursively up for a vcpkg-configuration.json).

strega-nil committed:
[vcpkg] virtualize VisitorCrtpBase

This allows us to get better compiler errors when playing around with json parsing
Additionally, move some stuff over to json.h, and some minor json reading changes
Also rename *Field -> *Deserializer
Additionally, move the valid fields check into the Reader

strega-nil committed:
[vcpkg] Add initial registries support

Also add format_manifest support for configuration

strega-nil committed:
[vcpkg cmake] don't overflow command line length

for format

strega-nil committed:
[vcpkg] finish off initial registries support

still to do: add other types of registries!

strega-nil committed:
[vcpkg] Move over to using registries for the builtin ports directory

this still allows us to use overlay ports, but uses the registry set for everything else

strega-nil committed:
[vcpkg] Add filesystem registries!

Edhebi commented:
In your example project, you specify both your local registry and the builtin one, with no default. Is it possible to only define a registry for some packages, and keep the builtin as a fallback ?

BillyONeal commented on toolsrc/src/vcpkg/paragraphs.cpp:
```diff
@@ -343,34 +344,71 @@ namespace vcpkg::Paragraphs
+            for (auto&& path : port_dirs)
+            {
+                auto port_name = fs::u8string(path.filename());
+                if (packages.size() == 0 || std::find(packages.begin(), packages.end(), port_name) != packages.end())
```
Something seems a little screwy here: we filter by packages which presumably the registry already knows? It seems like some comments as to what this is doing would help.

Maybe this should be called 'filter' rather than packages?

strega-nil replied:
Yeah, this doesn't make a lot of sense; I think it should just push all of the directory names.

BillyONeal commented on toolsrc/src/vcpkg/paragraphs.cpp:
```diff
@@ -343,34 +344,71 @@ namespace vcpkg::Paragraphs
-            auto maybe_spgh = try_load_port(fs, path);
+            load_port_names(registry.implementation(), registry.packages());
+        }
+        if (auto registry = registries.default_registry())
```
Why is the default registry a special snowflake rather than just being the last registry in registries() ?

strega-nil replied:
Because the default registry is a special snowflake -- it can be null, there's only one of them, etc.

BillyONeal commented on toolsrc/include/vcpkg/vcpkgpaths.h:
```diff
@@ -44,11 +45,47 @@ namespace vcpkg
+        std::vector<Registry> registries_;
+    };
+
+    struct Configuration
```
Considering it's just a member of `vcpkgpaths` I think we should just add this to vcpkgpaths instead of introducing a wrapper struct. At such time we want to add more stuff we can always introduce a wrapper struct but "YAGNI" usually applies to this sort of thing.

strega-nil replied:
We have immediate plans to extend it; otherwise, I'd agree.

BillyONeal replied:
I see, this is called "`<configuration>`" in the RFC. In that case I think it should be called `RegistryConfiguration` or something like that. With this name it isn't clear, it might be a CMake configuration or similar.

BillyONeal replied:
I see, after further discussion I am OK with this.

BillyONeal commented on toolsrc/src/vcpkg/portfileprovider.cpp:
```diff
@@ -92,30 +94,62 @@ namespace vcpkg::PortFileProvider
+            auto registry_root = registry->get_registry_root(paths);
+            auto port_directory = registry_root / fs::u8path(spec);
+
+            if (fs.exists(port_directory))
```
Why do we block on this 'exists'? That seems like it should be a hard error for a registry to claim it has a thing but not have the thing.

strega-nil replied:
Let us consider the following layout:

r1: is registry for packages: [ "zlib" ]

```
boost
zlib
```

r2: default registry

```
meow
```

even tho we find the name `boost` in some registry, the registry that would find it isn't the owner of the package name. Thus, `boost` is not a loadable port, and so it is not considered as part of the set of all ports.

BillyONeal commented on toolsrc/src/vcpkg/portfileprovider.cpp:
```diff
@@ -129,20 +163,19 @@ namespace vcpkg::PortFileProvider
         cache.clear();
```
Well at least we are consistent in the 'letting pointers escape' 😥

This design problem is preexisting but it's super scary that this function is const but totally invalidates results from all previous calls to the function.

BillyONeal replied:
This is preexisting so I guess 'no change requested' but dear god that hurts to type here

ras0219 replied:
Yeah, this function should not be const.

BillyONeal commented on toolsrc/src/vcpkg/registries.cpp:
```diff
@@ -0,0 +1,108 @@
+        r.required_object_field(
+            type_name(), obj, KIND, kind, Json::StringDeserializer{"a registry implementation kind"});
+
+        if (kind == KIND_BUILTIN)
```
Stupid question: If we need an if/elseif ladder like this should we really use polymorphism for the different kinds?

strega-nil replied:
Yes, this is deserialization, not usage.

BillyONeal commented on toolsrc/src/vcpkg/registries.cpp:
```diff
@@ -0,0 +1,108 @@
+        }
+    }
+
+    StringView RegistryDeserializer::type_name() const { return "a registry"; }
```
Should it return StringLiteral rather than StringView?

Ditto elsewhere

strega-nil replied:
No, since it's a virtual override.

ras0219 replied:
```suggestion
    StringView RegistryDeserializer::type_name() const override { return "a registry"; }
```

strega-nil commented on toolsrc/include/vcpkg/base/util.h:
```diff
@@ -63,7 +65,10 @@ namespace vcpkg::Util
+    using FmapRefOut = decltype(std::declval<Func&>()(*std::declval<Range>().begin()));
+
+    template<class Range, class Func>
+    using FmapOut = std::decay_t<FmapRefOut<Range, Func>>;
```
A function can return a reference to an array, tho --

```
const int (&x)()[3] {
}
```

for example

strega-nil commented on toolsrc/include/vcpkg/registries.h:
```diff
@@ -0,0 +1,69 @@
+
+    struct RegistryImplDeserializer : Json::IDeserializer<std::unique_ptr<RegistryImpl>>
+    {
+        constexpr static StringLiteral KIND = "kind";
+        constexpr static StringLiteral PATH = "path";
+
+        constexpr static StringLiteral KIND_BUILTIN = "builtin";
+        constexpr static StringLiteral KIND_DIRECTORY = "directory";
```
Better to avoid magic constants, IMO, since it's referred to multiple times. I'd rather a typo be a compile time error than a runtime error, with only a single place to screw that up.

strega-nil commented on toolsrc/include/vcpkg/vcpkgpaths.h:
```diff
@@ -44,11 +45,47 @@ namespace vcpkg
     struct PackageSpec;
     struct Triplet;
 
+    struct Registries
```
I originally called it RegistrySet; maybe that's better.

strega-nil commented on toolsrc/src/vcpkg/base/json.cpp:
```diff
@@ -1002,6 +1031,26 @@ namespace vcpkg::Json
         }
     }
 
+    std::pair<Value, JsonStyle> parse_file(vcpkg::LineInfo linfo,
```
We generally have overloads in vcpkg for "write to error code if a call fails", and additionally "fail fast if a call fails".

strega-nil commented on toolsrc/src/vcpkg/base/json.cpp:
```diff
@@ -1002,6 +1031,26 @@ namespace vcpkg::Json
+        if (ec)
+        {
+            System::print2(System::Color::error, "Failed to read ", fs::u8string(path), ": ", ec.message(), "\n");
+            Checks::exit_fail(linfo);
```
Yes, it's more interesting to get who called the function.

strega-nil commented:
@Edhebi yep! The default registry defaults to the builtin one :) I'd definitely recommend reading the manifest RFC at #12881 for more info, and if you still have questions post them there :)

ras0219 commented on toolsrc/src/vcpkg/commands.format-manifest.cpp:
```diff
@@ -41,16 +43,25 @@ namespace
             return nullopt;
         }
 
+        Optional<Json::Value> config;
+        if (auto conf = parsed_json_obj.get("configuration"))
+        {
+            config = std::move(*conf);
+        }
+
```
Based on my reading, it saves a copy of the configuration json object if the declaration of `parsed_json` is changed to be `auto&`.

ras0219 commented on toolsrc/include/vcpkg/base/optional.h:
```diff
@@ -162,11 +165,12 @@ namespace vcpkg
-        template<class T, bool B>
-        struct OptionalStorage<T&, B>
+        template<class T>
+        struct OptionalStorage<T&, true>
```
Yeah, I made this change without considering that case. Seems trivial to reverse it.

ras0219 commented on toolsrc/include/vcpkg/portfileprovider.h:
```diff
@@ -27,13 +27,13 @@ namespace vcpkg::PortFileProvider
     {
         explicit PathsPortFileProvider(const vcpkg::VcpkgPaths& paths,
-                                       const std::vector<std::string>& ports_dirs_paths);
+                                       const std::vector<std::string>& overlay_ports /*, int finder */);
```
```suggestion
                                       const std::vector<std::string>& overlay_ports);
```

strega-nil committed:
Merge remote-tracking branch 'upstream/master' into filesystem-registries

strega-nil committed:
CRs

strega-nil committed:
format

strega-nil committed:
add support for classic mode

strega-nil committed:
format

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into filesystem-registries

strega-nil committed:
fix shadowing warning

BillyONeal commented on toolsrc/include/vcpkg/vcpkgpaths.h:
```diff
@@ -45,10 +49,41 @@ namespace vcpkg
+
+    struct Configuration
+    {
+        RegistrySet registries;
```
The spec says that default-registry is a property of the Configuration but here it's a property of the RegistrySet. That's OK (since I expect in most cases we'll need those together) but maybe there should be a comment:

```
// default-registry is part of `registries`
```
or similar?

BillyONeal commented on toolsrc/src/vcpkg/paragraphs.cpp:
```diff
@@ -343,34 +344,67 @@ namespace vcpkg::Paragraphs
+            auto port_dirs = fs.get_files_non_recursive(r.get_registry_root(paths));
+            Util::sort(port_dirs);
+            Util::erase_remove_if(port_dirs, [&](auto&& port_dir_entry) {
+                return fs.is_regular_file(port_dir_entry) && port_dir_entry.filename() == ".DS_Store";
```
Note: the `b-/boost` resolution in the spec is missing here. Add a TODO?

BillyONeal replied:
(And/or should this filter for our valid port identifier names instead of looking for .DS_Store. Yes I know this code is preexisting so no need to touch that yet)

BillyONeal commented on toolsrc/src/vcpkg/base/files.cpp:
```diff
@@ -1159,6 +1158,10 @@ namespace vcpkg::Files
         {
             return rhs;
         }
+        else if (!rhs.root_directory().empty())
```
This seems like it is unit testable.

BillyONeal commented on toolsrc/src/vcpkg/paragraphs.cpp:
```diff
@@ -343,34 +344,67 @@ namespace vcpkg::Paragraphs
+
+        std::vector<std::string> ports;
+
+        auto load_port_names = [&fs, &ports, &paths](const RegistryImpl& r) {
```
Proposal: Extract a function here:

```
void append_port_names(std::vector<std::string>& ports, const VcpkgPath& paths, const path& registryRoot)
```

It seems the closure is adding a lot of shared mutable state which makes proving correctness on a function by function basis more difficult for no gain here. Moreover, this whole block is repeated for overlay ports on line 440 whereas if it were a function instead of a lambda that code could be shared.

BillyONeal commented on toolsrc/src/vcpkg/portfileprovider.cpp:
```diff
@@ -92,30 +94,62 @@ namespace vcpkg::PortFileProvider
             {
-                auto found_scf = Paragraphs::try_load_port(filesystem, ports_spec);
+                auto found_scf = Paragraphs::try_load_port(fs, ports_spec);
+                if (auto scf = found_scf.get())
```
If you want you could save an indent level by flipping this condition since the else block terminates the program. On the other hand, I like the positive if so up to you.

BillyONeal commented on toolsrc/src/vcpkg/vcpkgpaths.cpp:
```diff
@@ -81,6 +83,230 @@ namespace
 
 namespace vcpkg
 {
+    RegistrySet::RegistrySet() : default_registry_(Registry::builtin_registry()), registries_() { }
```
Hmmm should these be in `registries.cpp`?

strega-nil replied:
RegistrySet was living in vcpkgpaths, but yeah, perhaps.

strega-nil replied:
Oh right, the reason for this is that `Configuration` holds a `RegistrySet` value, so it needs to see the definition of `RegistrySet`.

BillyONeal commented on toolsrc/src/vcpkg/vcpkgpaths.cpp:
```diff
@@ -81,6 +83,230 @@ namespace
+        return std::move(parsed_config_opt).value_or_exit(VCPKG_LINE_INFO);
+    }
+
+    static std::pair<Optional<std::pair<Json::Object, Json::JsonStyle>>, std::pair<fs::path, Configuration>>
```
Pairs of pairs really sound like we should make up a type here :)

BillyONeal commented on toolsrc/src/vcpkg/vcpkgpaths.cpp:
```diff
@@ -81,6 +83,230 @@ namespace
+                config_obj = std::move(p_manifest_config->object());
+                config_dir = manifest_dir;
+            }
+            else if (p_manifest_config->is_string())
```
Is this behavior something we have some way to test right now in the "integration tests"? (It's OK if the answer is no)

BillyONeal reviewed (approved):
No serious issues left other than Bill not being a fan of that one lambda. I don't agree with the "configuration" mechanism in the spec, but this does seem to implement what's in that spec.

strega-nil committed:
billy crs

strega-nil committed:
fix combine on linux

strega-nil committed:
Change how vcpkg-configuration works after discussion with team

We had a serious discussion today, and so the way we're going forward
with this is very different -- no longer having a `"configuration"`
field, adding inheritance, etc.

Check the RFC for more info on how this will be done in the future; for
the MVP, all it changes (fundamentally) is where the configuration stuff
gets set -- it's in vcpkg-configuration.json in the manifest root.

strega-nil committed:
formatting

strega-nil commented:
@BillyONeal could you merge this today?

BillyONeal commented:
Yes, but I thought we wanted another reviewer?

ras0219 reviewed (commented):
Please also confirm everything still compiles using v140 and gcc-7:
```Dockerfile
FROM ubuntu:18.04

RUN apt-get update && DEBIAN_FRONTEND=noninteractive apt-get install -y \
  xz-utils \
  build-essential \
  tar \
  unzip \
  ninja-build \
  cmake \
  curl \
  vim \
  git \
  nasm \
  yasm \
  python3 \
  pkg-config \
  zip

CMD [ "/bin/bash" ]
```
>docker run -it --rm 51d7a08f83a8 /bin/bash -c "git clone https://github.com/microsoft/vcpkg --depth=1 && vcpkg/bootstrap-vcpkg.sh"

```json
// CMakeSettings.json
[
    {
      "name": "x64-Release-v140",
      "generator": "Visual Studio 16 2019 Win64",
      "configurationType": "RelWithDebInfo",
      "buildRoot": "${projectDir}\\out\\build\\${name}",
      "installRoot": "${projectDir}\\out\\install\\${name}",
      "cmakeCommandArgs": "-T v140 -DVCPKG_STANDARD_LIBRARY=msvc-stl -DVCPKG_WARNINGS_AS_ERRORS=OFF -DBUILD_TESTING=OFF",
      "buildCommandArgs": "/m",
      "ctestCommandArgs": "",
      "inheritEnvironments": [ "msvc_x64_x64" ],
      "variables": [],
      "environments": [ { "CL": "/MP" } ]
    }
]
```

strega-nil commented on toolsrc/src/vcpkg/sourceparagraph.cpp:
```diff
@@ -479,202 +357,136 @@ namespace vcpkg
-        StringView type_name() { return "a package name"; }
+        virtual StringView type_name() const override { return "a feature"; }
+
+        constexpr static StringLiteral NAME = "name";
+        constexpr static StringLiteral DESCRIPTION = "description";
+        constexpr static StringLiteral DEPENDENCIES = "dependencies";
```
This, I will disagree with. `constexpr static` is the right way around.

BillyONeal replied:
I think the point is that we should pick one and stick with it rather than arguing for one or the other.

strega-nil commented:
@ras0219 fwiw, we compile with g++-7 in CI.

Also, I'll have a patch out later today to fix compilation with g++-6.

strega-nil commented on toolsrc/include/vcpkg/registries.h:
```diff
@@ -0,0 +1,105 @@
+    {
+        RegistrySet();
+
+        friend void swap(RegistrySet& lhs, RegistrySet& rhs) noexcept
```
That's a good point... I feel like std::swap is fine for us here.

strega-nil commented on toolsrc/src/vcpkg/configuration.cpp:
```diff
@@ -0,0 +1,66 @@
+#include <vcpkg/base/system.print.h>
```
This response doesn't make sense to me; configuration is going to get added to. We could make `registries.cpp` part of `configuration.cpp`, I guess, but it's not like it's going to change the compilation time that much.

strega-nil commented on toolsrc/include/vcpkg/registries.h:
```diff
@@ -0,0 +1,105 @@
+{
+    struct RegistryImpl
+    {
+        virtual void update(VcpkgPaths& paths, std::error_code& ec) const = 0;
```
We can worry about this later.

strega-nil commented on toolsrc/include/vcpkg/registries.h:
```diff
@@ -0,0 +1,105 @@
+
+    struct RegistryImplDeserializer : Json::IDeserializer<std::unique_ptr<RegistryImpl>>
+    {
+        constexpr static StringLiteral KIND = "kind";
+        constexpr static StringLiteral PATH = "path";
+
+        constexpr static StringLiteral KIND_BUILTIN = "builtin";
+        constexpr static StringLiteral KIND_DIRECTORY = "directory";
```
Acknowledged, but I disagree. I prefer putting them in the type.

strega-nil committed:
Robert's CRs

BillyONeal commented:
Thanks for your contribution!
